src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,347 @@
+use std::convert::TryInto;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use chrono::{NaiveDate, NaiveDateTime, NaiveTime};
+use futures::stream::BoxStream;
+use futures::{pin_mut, StreamExt};
+use genawaiter::sync::{Co, Gen};
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use postgres_protocol::message::backend::ReplicationMessage;
+use postgres_protocol::replication::pgoutput::{
+    LogicalReplicationMessage, PgOutput, Tuple, TupleData,
+};
+use tokio_postgres::binary_copy::BinaryCopyOutStream;
+use tokio_postgres::replication_client::{ReplicationClient, SnapshotMode};
+use tokio_postgres::types::{PgLsn, Type as PgType};
+use tokio_postgres::{connect_replication, NoTls, ReplicationMode};
+use uuid::Uuid;
+
+use crate::source::{SimpleSource, StreamItem, Timestamper};
+use dataflow_types::PostgresSourceConnector;
+use expr::parse_datum;
+use repr::{Datum, Row, RowPacker, ScalarType};
+
+lazy_static! {
+    /// Postgres epoch starts at 2000-01-01T00:00:00Z
+    static ref PG_EPOCH: SystemTime = UNIX_EPOCH + Duration::from_secs(946_684_800);
+}
+
+/// Information required to sync data from Postgres
+pub struct PostgresSimpleSource {
+    connector: PostgresSourceConnector,
+    packer: RowPacker,
+    plugin: PgOutput,
+}
+
+struct PgColumn {
+    scalar_type: PgType,
+    nullable: bool,
+}
+
+impl PostgresSimpleSource {
+    /// Constructs a new instance
+    pub fn new(connector: PostgresSourceConnector) -> Self {
+        let publication = connector.publication.clone();
+        Self {
+            connector,
+            packer: RowPacker::new(),
+            plugin: PgOutput::new(vec![publication]),
+        }
+    }
+
+    /// Uses a normal postgres connection (i.e not in replication mode) to get the column types of
+    /// the remote table as well as the relation id that is used to filter the replication stream
+    async fn table_info(&self) -> (u32, Vec<PgColumn>) {
+        let conninfo = &self.connector.conn;
+
+        let (client, connection) = tokio_postgres::connect(&conninfo, NoTls).await.unwrap();
+
+        // TODO communicate errors back into the stream
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                eprintln!("connection error: {}", e);
+            }
+        });
+
+        let namespace = &self.connector.namespace;
+        let table = &self.connector.table;
+        let rel_id: u32 = client
+            .query(
+                "SELECT c.oid
+                    FROM pg_catalog.pg_class c
+                    INNER JOIN pg_catalog.pg_namespace n
+                          ON (c.relnamespace = n.oid)
+                    WHERE n.nspname = $1
+                      AND c.relname = $2;",
```

Comment:
Can you explain what this query is doing?

Reply:
Everything has an OID in postgres. This query gets the OID of the requested table. I think you can get the same result with:

```
SELECT 'namespace.table'::regclass::oid;
```

Reply:
Hah, I didn't know about this syntax. It's much simpler, thank you!

I took this from upstream that I guess uses this long form to take the extra info about replica identity and relkind https://github.com/postgres/postgres/blob/REL_13_2/src/backend/replication/logical/tablesync.c#L650-L659